### PR TITLE
New API

### PR DESCRIPTION
**Pull request by mmarkelov**
Thinking about rewriting **jest-playwright** with new API:
 Using ```jest-playwright --parallel``` command is kind of hack, so I think about replacing it.

To run tests for specified browsers in you config you should do:
**jest-playwright.config.js**:
```js
module.exports = {
    browsers: [ 'firefox', 'webkit', 'chromium'],
}
```

In tests:
```js
describe('Google', () => {
    beforeAll(async () => {
       // It will navigate all browsers to url
        await page.goto('https://whatismybrowser.com/')
    })

    it('should display "google" text on page', async () => {
        const titles = await page.title()
        // Thinking about more clear and nice way to handle it
        titles.forEach(title => expect(title).toBe('What browser am I using? Is my browser up to date?'))
       // It will just return result for all browsers in array
        const [firefox, webkit, chrome] = await page.$eval('.string-major', el => el.innerHTML)

        expect(firefox).toContain('Firefox')
        expect(webkit).toContain('Safari')
        expect(chrome).toContain('Chrome')

    })
})
```

Let me know what do you think about it?

### CODE REVIEW

**Comment by mxschmitt** [comment]
Thats a nice idea! So you are just replacing the implementation for multiple browsers as far as I've seen right?
Not fully sure about the `const [firefox, webkit, chrome]` part, maybe replace it with an object, because then it's easier to handle which key is associated with with browser by using object deconstruction.
Also your solution would mean in the end, that the user always have to check for `N` custom browsers which might be not so good for all the users and from the clean coding perspective.
But for every solution there are some disadvantages and when we compare this solution with the previous one, then this one is much nicer from my perspective.

**Comment by mmarkelov** [comment]
@mxschmitt yeah! That is what I'm thinking about.
Replacing ```const [firefox, webkit, chrome]``` is a good idea, I suppose.
Yes, it will be not the cleanest way to do tests. But I think it is better way to tests multiple browsers.
Of cause this is not perfect, because of this:
```js
    it('should display "google" text on page', async () => {
        const titles = await page.title()
        // That is not the best way
        titles.forEach(title => expect(title).toBe('What browser am I using? Is my browser up to date?'))
       // It will just return result for all browsers in array
        const {firefox, webkit, chrome} = await page.$eval('.string-major', el => el.innerHTML)
    })
```

I'm trying to improve it, but not sure that it will be possible to simplify it more

**Comment by mxschmitt** [comment]
> @mxschmitt yeah! That is what I'm thinking about.
> Replacing `const [firefox, webkit, chrome]` is a good idea, I suppose.
> Yes, it will be not the cleanest way to do tests. But I think it is better way to tests multiple browsers.
> Of cause this is not perfect, because of this:
> 
> ```js
>     it('should display "google" text on page', async () => {
>         const titles = await page.title()
>         // That is not the best way
>         titles.forEach(title => expect(title).toBe('What browser am I using? Is my browser up to date?'))
>        // It will just return result for all browsers in array
>         const {firefox, webkit, chrome} = await page.$eval('.string-major', el => el.innerHTML)
>     })
> ```
> 
> I'm trying to improve it, but not sure that it will be possible to simplify it more

One Idea would be to force the user that he needs to make
```js
["chromium", "firefox", "webkit"].forEach(browserName => {
  const page = jestPlaywright.getBrowser(browserName)
  describe(`Testing browser ${browserName}`, () => {
    it("should do XYZ", async() => {
      
    })
  })
})
```
idk. Then reporting would also be kinda better afaik.

**Comment by mmarkelov** [comment]
Some new thoughts here. I just added helper proxy to simplify test process:
```js
describe('Google', () => {
  beforeAll(async () => {
    // It will navigate all browsers to url
    await page.goto('https://whatismybrowser.com/');
  });

  it('should display "google" text on page', async () => {
    const titles = await page.title();
    // It will expect toBe for all browsers

    // Passed tests
    expectAllBrowsers(titles).toBe('What browser am I using? Is my browser up to date?')

    // Failed tests
    const myBrowser = await page.$eval(
      '.string-major',
      el => el.innerHTML,
    );

    expectAllBrowsers(myBrowser).toContain('Chrome')
  });
});
```

I just think about nice error message here. For now it will be:
```
 FAIL  ./index.test.js (5.459s)
  Google
    ✕ should display "google" text on page (104ms)

  ● Google › should display "google" text on page

    expect(received).toContain(expected) // indexOf

    Expected substring: "Chrome"
    Received string:    "·······························
                                    <a href=\"/detect/what-version-of-firefox-do-i-have?utm_source=whatismybrowsercom&amp;utm_medium=internal&amp;utm_campaign=homepage&amp;utm_content=simple-major\">Firefox 73 on macOS (Catalina)</a>·······························
                    "

      18 |     );
      19 | 
    > 20 |     expectAllBrowsers(myBrowser).toContain('Chrome')
         |                                  ^
      21 |   });
      22 | });
      23 | 

      at browsers.forEach.browser (node_modules/jest-playwright-preset/lib/PlaywrightEnvironment.js:112:67)
          at Array.forEach (<anonymous>)
      at Proxy.toContain (node_modules/jest-playwright-preset/lib/PlaywrightEnvironment.js:107:34)
      at Object.it (index.test.js:20:34)

Failed test for firefox
```

**Comment by mxschmitt** [comment]
cool! let's maybe ask the users who are actually using this feature what they think about that before we merge?

**Comment by mmarkelov** [comment]
@mxschmitt yeah! I just want to support ```jest-playwright``` commands for some time. But I want to replace it with this one soon

File: src/PlaywrightEnvironment.ts [hunk 1]
```diff
@@ -1,6 +1,9 @@
 /* eslint-disable no-console */
 import NodeEnvironment from 'jest-environment-node'
 import { Config as JestConfig } from '@jest/types'
+import Expect = jest.Expect
+import playwright, { Browser, BrowserContext, Page } from 'playwright'
```
Comment (mxschmitt) [inline src/PlaywrightEnvironment.ts:5]: please import types from playwright-core

**Comment by mmarkelov** [comment]
Some updates here. Just tried to use it with **jest-circus** runner, it is ok. But for some reason it needs to run with **forceExit** and **detectOpenHandles** options. Still working on it
```js
jest --forceExit --detectOpenHandles
```

**Comment by mmarkelov** [comment]
Some updates. I got passed tests on local machine. But got some issues with `webkit`. But for now it's quite working solution. I think it's better to use tests like so:
```js
import path from 'path'

jest.setTimeout(10000)

describe('Example HTML file', () => {
  it('should detect the heading "Example" on page', async () => {
    await page.goto(`file:${path.join(__dirname, 'example.html')}`)
    const browser = await page.$eval('h1', (el) => el.textContent)
    expectAllBrowsers(browser).toBe('Example')
  })
})
```

Also I'm thinking about adding `expectChromium` and others for running **expect** function for specified browser. IDK how it will be relate to #96

File: jest-playwright.config.js [hunk 2]
```diff
@@ -0,0 +1,4 @@
+module.exports = {
+  USE_NEW_API: true,
```
Comment (mxschmitt) [inline jest-playwright.config.js:2]: why no camel case here?

File: src/PlaywrightEnvironment.ts [hunk 3]
```diff
@@ -1,16 +1,29 @@
 /* eslint-disable no-console */
 import NodeEnvironment from 'jest-environment-node'
 import { Config as JestConfig } from '@jest/types'
+import Expect = jest.Expect
```
Comment (mxschmitt) [inline src/PlaywrightEnvironment.ts:4]:
```suggestion
```

File: src/PlaywrightEnvironment.ts [hunk 4]
```diff
@@ -90,15 +103,234 @@ class PlaywrightEnvironment extends NodeEnvironment {
...
+        browsers.map((browser) =>
+          getPlaywrightInstance(playwrightPackage, browser),
+        ),
+      )
+
+      // Helpers
+      const getResult = <T>(
+        data: T[],
+        instances: BrowserType[] | Array<keyof typeof playwright.devices>,
+      ) => {
```
Comment (mxschmitt) [inline src/PlaywrightEnvironment.ts:140]:
```suggestion
      ): any => {
```

File: src/PlaywrightEnvironment.ts [hunk 5]
```diff
@@ -90,15 +103,234 @@ class PlaywrightEnvironment extends NodeEnvironment {
...
+        const result: any = {}
```
Comment (mxschmitt) [inline src/PlaywrightEnvironment.ts:141]:
```suggestion
        const result: any = {} // eslint-disable-line no-explicit-any
```

**Comment by mmarkelov** [comment]
> small nits. Would be cool to clean the ESLint warnings and use pragmas for the `any` cases. Then we can merge in my mind.

Yeah! I agree but it seems challenging for me) To fix whole TS warnings

File: jest-playwright.config.js (hunk 2 above)
Comment (mmarkelov) [reply jest-playwright.config.js:2]: @mxschmitt I think that snake style will show that it's experimental feature

**Comment by mxschmitt** [comment]
> > small nits. Would be cool to clean the ESLint warnings and use pragmas for the `any` cases. Then we can merge in my mind.
> 
> Yeah! I agree but it seems challenging for me) To fix whole TS warnings

Feel free to use `// eslint-disable-line` :)

**Comment by mmarkelov** [comment]
I'll close it, cause I suppose using custom jest-runner is better solution for running tests for multiple browsers support
#113

**Pull request closed without merging**